### PR TITLE
feat(contourSeg): polyline performance improvements

### PR DESCRIPTION
**Pull request by lscoder**
### Context

Adding contour annotations may result in polylines with thousands of points making the viewer much slower to render the annotations and also process them (append, remove, create holes, interpolate, etc.). 

### Changes & Results

Implemented the Ramer–Douglas–Peucker algorithm that simplifies polylines based on an a given distance `epsilon` reducing the amount of points in a polyline.

Two polylines with 6k (source) and 7k (target) points were used on the test below to measure the time it would take to merge them before and after simplifying the polylines.

perf_polyline-simplification_merge

A good way to generate contours with +50k points is zooming all the way out before creating a freehand segmentation contour. Both polylines got simplified to new polylines with ~200 points or less each with no visual changes on the viewport. I know this is not a real example but that helped to measure the time it would take to process the contours.

https://github.com/cornerstonejs/cornerstone3D/assets/931820/ce4e68b5-d714-479a-8113-52aaaefd6612

### Testing

Draw contours (spline, freehand, livewire) and make sure they are rendered and processed (append, remove, hole, interpolation, etc) correctly. A performance improvement is very noticeable when drawing multiple freehand ROIs.

### CODE REVIEW

File: packages/tools/src/utilities/math/polyline/simplify.ts [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+ * https://rosettacode.org/wiki/Ramer-Douglas-Peucker_line_simplification
+ * https://karthaus.nl/rdp/
+ *
+ * @param polyline - Polyline to simplify
+ * @param epsilon - A maximum given distance 'epsilon' to decide if a point
+ * should or shouldn't be added the the simplified polyline version. In each
+ * iteration the polyline is split into two polylines and the distance of each
+ * point from those new polylines are checked against the line that connects
+ * the first and last points.
+ * @returns Simplified polyline
```
Comment (sedghi) [inline packages/tools/src/utilities/math/polyline/simplify.ts:20]: is this our implementatoin? is the used code open permissive license?

File: packages/tools/src/utilities/contours/updateContourPolyline.ts [hunk 2]
```diff
@@ -14,6 +14,12 @@ import {
  * @param viewport - Viewport
  * @param polylineData - Polyline data (points, winding direction and closed)
  * @param transforms - Methods to convert points to/from canvas and world spaces
+ * @param options - Options
```
Comment (sedghi) [inline packages/tools/src/utilities/contours/updateContourPolyline.ts:17]: I don't see any tool using this option. I expected the simplify object being part of the tools configuration?

**Review by sedghi — COMMENTED** [review]
see my comments

File: packages/tools/src/utilities/math/polyline/simplify.ts (hunk 1 above)
Comment (lscoder) [reply packages/tools/src/utilities/math/polyline/simplify.ts:20]: This is 100% my implementation.

File: packages/tools/src/utilities/contours/updateContourPolyline.ts (hunk 2 above)
Comment (lscoder) [reply packages/tools/src/utilities/contours/updateContourPolyline.ts:17]: Updated spline, freehand and livewire configs

File: packages/tools/src/tools/annotation/SplineROITool.ts [hunk 3]
```diff
@@ -121,6 +120,18 @@ class SplineROITool extends ContourSegmentationBaseTool {
          * modifier must be pressed when the first point of a new contour is added.
          */
         contourHoleAdditionModifierKey: KeyboardBindings.Shift,
+        /**
+         * The polyline may get processed in order to reduce the number of points
+         * for better performance and storage.
+         */
+        decimate: {
+          enabled: true,
```
Comment (sedghi) [inline packages/tools/src/tools/annotation/SplineROITool.ts:128]: I prefer enabled to be false as default for now for all tools

File: packages/tools/src/tools/base/ContourBaseTool.ts [hunk 4]
```diff
@@ -210,6 +212,31 @@ abstract class ContourBaseTool extends AnnotationTool {
...
+      points: Types.Point2[];
+      closed?: boolean;
+      targetWindingDirection?: ContourWindingDirection;
+    },
+    transforms: {
+      canvasToWorld: (point: Types.Point2) => Types.Point3;
+    }
+  ) {
+    const decimateConfig = this.configuration?.decimate || {};
+    console.log('>>>>> decimate', {
```
Comment (sedghi) [inline packages/tools/src/tools/base/ContourBaseTool.ts:227]: remove

File: packages/tools/src/tools/annotation/SplineROITool.ts (hunk 3 above)
Comment (lscoder) [reply packages/tools/src/tools/annotation/SplineROITool.ts:128]: Updated

**Review by sedghi — APPROVED** [review]

**Pull request merged**